src/content/reference/device-os/firmware.md

```diff
@@ -1439,6 +1466,444 @@ Previously, there were overloads with a `ttl` (time-to-live) value. These have b
+ContentType instead.
+
+```cpp
+// PROTOTYPE
+CloudEvent& data(const char* data, size_t size);
+```
+
+#### data - from pointer and length with content type - CloudEvent
+
+Copies data from a buffer specified with pointer and length. This is typically used with binary data.
+
+See [Content-Type](#contenttype-cloudevent), above, for the valid values.
+
+```cpp
+// PROTOTYPE
+CloudEvent& data(const char* data, size_t size, ContentType type)
+```
+
+#### data - from String - CloudEvent
+
+Copies data from a `String` object.
+
+```cpp
+// PROTOTYPE
+CloudEvent& data(const String& data)
+```
+
+#### dataString - CloudEvent
+
+Returns a copy of the data in the `EventData` object as a `String`. This is typically used for text data.
+
+```cpp
+// PROTOTYPE
+String dataString() const
+```
+
+#### data from Buffer - CloudEvent
+
+Copies binary data from a `Buffer` object. This is an alternative to using a pointer and length.
+
+```cpp
+// PROTOTYPE
+CloudEvent& data(const Buffer& data)
+```
+
+#### data - to Buffer - CloudEvent
+
+Returns a copy of the data in the `CloudEvent` as a `Buffer` object. This is typically used for binary data.
+
+```cpp
+// PROTOTYPE
+Buffer data() const
+```
+
+#### data - from EventData - CloudEvent
+
+Copies data from an `EventData` object, which is a `Variant`. This is used for structured data.
+
+```cpp
+// PROTOTYPE
+CloudEvent& data(const EventData& data)
+```
+
+#### data - to EventData - CloudEvent
+
+Returns a copy of the structured data in the `CloudEvent` as an `EventData` object, which is a `Variant`.
+
+```cpp
+// PROTOTYPE
+EventData dataStructured() const
+```
+
+For more information, see [structured subscription](/reference/device-os/typed-publish/#structured-subscription) in the typed and extended publish page.
+
+### File read and write - CloudEvent
+
+Event data can be saved to a file and restored from a file easily. For more information about the file system, see [File system](/reference/device-os/file-system/) and [File system API](#file-system).
+
+#### loadData - CloudEvent
+
+Loads the data in a `CloudEvent` from a file. Note that this only loads the data, not the other parameters like name, content-type, etc.
+
+Returns 0 on success or a non-zero system error code.
+
+```cpp
+// PROTOTYPE
+CloudEvent& loadData(const char* path)
+```
+
+#### saveData - CloudEvent
+
+Saves the data in a `CloudEvent` to a file. Note that this only saves the data, not the other parameters like name, content-type, etc.
+
+Returns 0 on success or a non-zero system error code.
+
+```cpp
+// PROTOTYPE
+int saveData(const char* path)
+```
+
+### Utility methods - CloudEvent
+
+#### setSize - CloudEvent
+
+Set the size of the data. This is rarely needed.
+
+```cpp
+// PROTOTYPE
+int setSize(size_t size)
+```
+
+#### size - CloudEvent
+
+Get the size of the data. This is the encoded size so there is computation involved in getting this value. 
```

Comment:
Under the hood, the data is stored in an encoded form so the size is always known and doesn't require computation when queried via the underlying CoAP API.